File: 0355-design-twitter/0355-design-twitter.py
class Twitter:

    def __init__(self):
        self.count = 0
        self.following = defaultdict(set)
        self.tweet = defaultdict(list)

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweet[userId].append((self.count, tweetId))
        self.count -= 1

    def getNewsFeed(self, userId: int) -> List[int]:
        res = []

        minHeap = []

        # Include user's tweets
        if userId in self.tweet:
            index = len(self.tweet[userId]) - 1
            if index >= 0:  # Ensure index is valid
                count, tweet = self.tweet[userId][index]
                heapq.heappush(minHeap, (count, userId, index - 1, tweet))

        # Include tweets from followed users
        for followerId in self.following[userId]:
            if followerId in self.tweet and len(self.tweet[followerId]) > 0:
                index = len(self.tweet[followerId]) - 1
                if index >= 0:  # Ensure index is valid
                    count, tweet = self.tweet[followerId][index]
                    heapq.heappush(minHeap, (count, followerId, index - 1, tweet))

        # Extract up to 10 most recent tweets
        while minHeap and len(res) < 10:
            count, followerId, index, tweet = heapq.heappop(minHeap)
            res.append(tweet)
            if index >= 0:
                time, tweetId = self.tweet[followerId][index]
                heapq.heappush(
                    minHeap, (time, followerId, index - 1, tweetId)
                )  # Correct the order of arguments

        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        self.following[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.following[followerId]:
            self.following[followerId].remove(followeeId)


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: 0355-design-twitter/0355-design-twitter.js
var Twitter = function() {
    this.time = 0;
    this.tweet = new Map();
    this.following = new Map();
};

Twitter.prototype.postTweet = function(userId, tweetId) {
    if (!this.tweet.has(userId)) {
        this.tweet.set(userId, []);
    }
    this.tweet.get(userId).push({ tweetId, time: this.time++ });
};

Twitter.prototype.getNewsFeed = function(userId) {
    let res = [];
    
    if (!this.following.has(userId)) {
        this.following.set(userId, new Set());
    }
    const users = this.following.get(userId);
    users.add(userId); // Add the user themselves to get their tweets
    
    let allTweets = [];
    for (let user of users) {
        const userTweets = this.tweet.get(user) || [];
        allTweets = allTweets.concat(userTweets);
    }
    
    // Sort all tweets by time in descending order
    allTweets.sort((a, b) => b.time - a.time);
    
    // Get the 10 most recent tweets
    for (let i = 0; i < Math.min(10, allTweets.length); i++) {
        res.push(allTweets[i].tweetId);
    }
    
    return res;
};

Twitter.prototype.follow = function(followerId, followeeId) {
    if (!this.following.has(followerId)) {
        this.following.set(followerId, new Set());
    }
    this.following.get(followerId).add(followeeId);
};

Twitter.prototype.unfollow = function(followerId, followeeId) {
    if (this.following.has(followerId)) {
        this.following.get(followerId).delete(followeeId);
    }
};